### DISCUSSION TITLE
NEWBIE question: How can I visualise a fitted timeseries model on my training data?

### DISCUSSION THREAD

**Discussion by heinerlehr** [autogluon/autogluon#4188] [Q&A]
Sorry if this is obvious - it isn't to me.

I would like to visualise the fit that I get from fit() on the training data for timeseries. I can forecat and show the forecast, I can include in the forecast training data, but I can't seem to find an option to include the fit to the training data. I was trying to use predict to "predict" both part of the training data set as well as generate a forecast, but it doesn't generate predictions for the training set (or I don't know how). Can anyone help, please?

**Comment by shchur** [comment]
Hi, thank you for the question. `TimeSeriesPredictor.predict(data)` always returns the forecast for the next `prediction_length` time steps starting from the end of the time series in `data`. 

Currently, we don't provide an option to generate in-sample (=historic/training) one-step-ahead predictions since many models in AutoGluon do not offer this functionality. For example, sequence-to-sequence (seq2seq) models like TemporalFusionTransformer, PatchTST, Chronos always return a prediction for the future `prediction_length` time steps, so it's computationally rather expensive to extract these one-step-ahead historic forecasts from them.

In summary, the only way to get these predictions from the current API is to call `predict` multiple times, something like
```python
shortest_ts_length = train_data.num_timesteps_per_item().min()
in_sample_predictions = []
for i in range(10, shortest_ts_length):  # ignore the first 10 observations since they are hard to predict
    sliced_data = train_data.slice_by_timestep(None, i)
    predictions_for_step_i = predictor.predict(sliced_data).slice_by_timestep(0, 1)  # select the 1-step-ahead predictions
    in_sample_predictions.append(predictions_for_step_i)
in_sample_predictions = pd.concat(in_sample_predictions)
```

**Comment by heinerlehr** [comment]
Dear Oleksandr,
Thank you for taking the time to respond - at least it wasn't obvious... Thanks also very much for providing me with a way to get what I want.
Kudos to you and your colleagues for the great work.
Heiner

**Comment by georgejosephada** [comment]
Of course you can (autogluon v1.3.1):

```
import numpy as np
import pandas as pd
from autogluon.timeseries import TimeSeriesDataFrame, TimeSeriesPredictor
import warnings
import os
import pickle
import shutil

shutil.rmtree("./autogluon_ensemble_test", ignore_errors=True)

# Suppress common warnings for a cleaner output
warnings.filterwarnings("ignore", "path already exists!.*", UserWarning)

# --- 1. Create Bogus Data ---
print("Step 1: Creating bogus time series data...")
FH = 3
VAL_WIN = 10
N_SAMPLES = 10_000
item_ids = ["A", "B"]
timestamps = pd.to_datetime(pd.date_range(start="2000-01-01", periods=N_SAMPLES, freq="D"))
data_list = []
for i, item in enumerate(item_ids):
    trend = np.arange(N_SAMPLES) * (i + 1) * 0.1
    seasonal = np.sin(np.arange(N_SAMPLES) / (FH + i)) * VAL_WIN * (i + 1)
    noise = np.random.normal(0, 1, N_SAMPLES)
    target_values = N_SAMPLES//2 + trend + seasonal + noise
    df = pd.DataFrame({"item_id": item, "timestamp": timestamps, "target": target_values, "past_covariate": target_values*0.1})
    data_list.append(df)
train_data = TimeSeriesDataFrame(pd.concat(data_list))
display(train_data.describe())
print("Bogus data created successfully.")


# --- 2. Train a Predictor ---
print("\nStep 2: Training a predictor...")
predictor = TimeSeriesPredictor(
    prediction_length=FH,
    path="./autogluon_ensemble_test",
    eval_metric="RMSSE",
    quantile_levels=[0.1, 0.9],
    cache_predictions=True
).fit(
    train_data,
    presets="best_quality",
    num_val_windows=VAL_WIN,
    enable_ensemble=True,
    time_limit=60,
)

# --- 3. Retrieve Historical Forecast for the Best Model ---
print("\nStep 3: Retrieving historical forecast for the best model...")
best_model_name = predictor.model_best
historical_forecast_df = None  # Initialize variable

print(f"Best model is '{best_model_name}'.")

# CASE 1: The best model IS an ensemble
if "Ensemble" in best_model_name:
    print("Best model is an ensemble. Blending base model OOF predictions...")
    model_info = predictor.info()['model_info']
    weights = model_info.get(best_model_name, {}).get('model_weights', {})
    base_models = list(weights.keys())
    print(f"Found ensemble weights: {weights}")

    all_oof_dfs = {}
    print("\nLoading OOF predictions for each base model...")
    for model in base_models:
        try:
            oof_path = os.path.join(predictor.path, "models", model, "utils", "oof.pkl")
            with open(oof_path, 'rb') as f:
                oof_prediction_list = pickle.load(f)
            all_oof_dfs[model] = pd.concat(oof_prediction_list)
            print(f"  > Loaded OOF for '{model}'")
        except Exception as e:
            print(f"  > ERROR loading OOF for '{model}': {e}")
            
    if all_oof_dfs:
        # Initialize an empty DataFrame to store the blended results
        template_df = next(iter(all_oof_dfs.values()))
        blended_df = pd.DataFrame(0.0, index=template_df.index, columns=template_df.columns)

        # Calculate the weighted average for each column (mean, 0.1, 0.9, ...)
        for model, weight in weights.items():
            if model in all_oof_dfs:
                blended_df += all_oof_dfs[model].fillna(0) * weight
        
        historical_forecast_df = blended_df

# CASE 2: The best model is NOT an ensemble
else:
    print(f"Best model is a base model. Loading its OOF predictions...")
    try:
        oof_path = os.path.join(predictor.path, "models", best_model_name, "utils", "oof.pkl")
        with open(oof_path, 'rb') as f:
            oof_prediction_list = pickle.load(f)
        
        # Combine all validation windows into one DataFrame
        historical_forecast_df = pd.concat(oof_prediction_list)
    except Exception as e:
        print(f"  > ERROR loading OOF for '{best_model_name}': {e}")

# Display the final result, regardless of which case was triggered
if historical_forecast_df is not None:
    print("\n--- ✅ Successfully Retrieved Historical (OOF) Forecast ---")
    display(historical_forecast_df.head())
else:
    print("\n--- ❌ Failed to retrieve historical forecast ---")
```
